### PR TITLE
Port `ser::Serializer` to `io::Write`

### LINKED ISSUE #202: Add function to serialize to a Writer
Bincode has [`serialize_into`](https://docs.rs/bincode/1.2.1/bincode/fn.serialize_into.html) which supports an arbitrary Writer, this would be very useful for RON as well, especially with `BufWriter`, because the data gets very large.

I'm seeing huge RAM spikes because of RON serialization: From ~90 MB baseline RAM usage to over 700 MB when it's serializing. Btw, the resulting ron file is 20 MB large.

image

---

In the above application, I'm serializing app state to disk every 10s.
Btw, the CPU usage is also quite high. This is the CPU & RAM usage when serializing to bincode instead. With ROM, the CPU usage is almost always at ~27% (probably because it takes so long), with bincode, it only shows short spikes (to 24%) when serializing:

image

---

Anyway, with a function like `serialize_into`+`BufWriter` we could at least reduce the RAM spikes.
This would also allow chaining it with e.g. a Gzip encoder for streaming compression, e.g. like with bincode:
```rust
	bincode::serialize_into(
		GzBuilder::new()
			.write(BufWriter::new(File::create(path)?), Compression::best()),
		&data,
	)?;
```

### PR DESCRIPTION
**Pull request by Plecra**
Makes the `Serializer` struct generic over `io::Write`, allowing it to be used with more types (and less allocation). As a side-effect, the trailing commas are also removed from the output.

This was a quick implementation, written in response to compiler messages, so it could probably do with some cleaning up. It might be a good time to approach #175 

Closes #202 

## TODO

- [ ] Update docs

### CODE REVIEW

File: src/ser/mod.rs [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
 mod value;
 
+/// Serializes `value` into `writer`
+pub fn to_writer<W, T>(writer: W, value: &T) -> Result<()>
+where
+    W: io::Write,
+    T: Serialize,
+{
+    let mut s = Serializer::new(writer, None, false)?;
+    value.serialize(&mut s)?;
```
Comment (kvark) [inline src/ser/mod.rs:19]: could just leave that without `?`?

File: src/ser/mod.rs [hunk 2]
```diff
@@ -176,10 +203,10 @@ fn default_depth_limit() -> usize {
 }
 
 fn default_new_line() -> String {
-    #[cfg(not(target_os = "windows"))]
+    // #[cfg(any(test, not(target_os = "windows")))]
```
Comment (kvark) [inline src/ser/mod.rs:206]: why are these changes needed?

File: src/ser/mod.rs [hunk 3]
```diff
@@ -214,29 +241,26 @@ impl Default for PrettyConfig {
 ///
 /// You can just use `to_string` for deserializing a value.
 /// If you want it pretty-printed, take a look at the `pretty` module.
-pub struct Serializer {
-    output: String,
+pub struct Serializer<W: io::Write> {
```
Comment (kvark) [inline src/ser/mod.rs:244]: this is great! I was doing the same in [Naga](https://github.com/gfx-rs/naga/blob/1a1f6a47b06c89abb1dff70326c076f1088964a3/src/back/msl.rs#L308) and thinking that it would be cool to get back to Ron and fix it :)

File: src/ser/mod.rs [hunk 4]
```diff
@@ -248,13 +272,13 @@ impl Serializer {
...
             is_empty: None,
-        }
+        })
     }
 
     /// Consumes `self` and returns the built `String`.
-    pub fn into_output_string(self) -> String {
-        self.output
-    }
+    // pub fn into_output_string(self) -> String {
```
Comment (kvark) [inline src/ser/mod.rs:279]: just remove this?

File: src/ser/mod.rs [hunk 5]
```diff
@@ -569,102 +617,151 @@ impl<'a> ser::Serializer for &'a mut Serializer {
...
-        self.output += "]";
+        ser.output.write(b"]")?;
         Ok(())
     }
 }
-
-impl<'a> ser::SerializeTuple for &'a mut Serializer {
+pub enum State {
+    First,
+    Rest
```
Comment (kvark) [inline src/ser/mod.rs:690]: where are your trailing commas?

File: tests/129_indexmap.rs [hunk 6]
```diff
@@ -16,7 +16,7 @@ tasks: {
         "-h",
       ]),
       ch_dir: Some("/")
-    ),
```
Comment (kvark) [inline tests/129_indexmap.rs:19]: please change this back to emit the comma in case the elements are not all on the same line

File: tests/depth_limit.rs [hunk 7]
```diff
@@ -26,12 +26,12 @@ struct Nested {
...
-    map: {8:'1',},
-    nested: (a:\"a\",b:'b',),
-    var: A(255,\"\",),
-    array: [(),(),(),],
+    float: (2.18,-1.1),
+    tuple: ((),false),
+    map: {8:'1'},
+    nested: (a:\"a\",b:'b'),
+    var: A(255,\"\"),
+    array: [(),(),()]
```
Comment (kvark) [inline tests/depth_limit.rs:34]: same here

File: src/ser/mod.rs (hunk 2 above)
Comment (Plecra) [reply src/ser/mod.rs:206]: Some of the tests expect linux line endings, so this was just a temporary fix to get them to pass. Didn't mean to push it 😄 Those tests should probably be fixed, though that's not relevant to the PR

Comment (Plecra) [reply src/ser/mod.rs:206]: Oh, looks like I did make it handle tests at build time actually. Would you like to implement this another way or just use the `any(test, ...` approach?

**Comment by kvark** [comment]
It's hard to look through your logic changes because of the formatting changes. Is there a way you could separate these in commits?

**Comment by Plecra** [comment]
In case you didn't get a notification, I've separated the commits

File: tests/depth_limit.rs [hunk 8]
```diff
@@ -55,5 +55,5 @@ fn depth_limit() {
         .with_new_line("\n".to_string());
     let s = ron::ser::to_string_pretty(&data, pretty);
 
-    assert_eq!(s, Ok(EXPECTED.to_string()));
+    assert_eq!(s.unwrap(), EXPECTED.to_string());
```
Comment (kvark) [inline tests/depth_limit.rs:58]: why this change?

File: tests/preserve_sequence_ex1.ron [hunk 9]
```diff
@@ -7,14 +7,14 @@
         3: '9',
         4: '1',
         5: '2',
-        6: '3',
+        6: '3'
```
Comment (kvark) [inline tests/preserve_sequence_ex1.ron]: This is not desired. Can we change the implementation to only omit the trailing comma for single-line things?

Comment (Plecra) [reply tests/preserve_sequence_ex1.ron]: I'm not sure what happened here, but I've already made this fix in the last batch of changes

File: tests/depth_limit.rs (hunk 8 above)
Comment (Plecra) [reply tests/depth_limit.rs]: I don't think there was any reason, and I've reverted it

Comment (Plecra) [reply tests/depth_limit.rs]: Oh I see... 

```sh
error[E0369]: binary operation `==` cannot be applied to type `std::result::Result<std::string::String, ron::ser::Error>`
  --> tests\depth_limit.rs:58:5
   |
58 |     assert_eq!(s, Ok(EXPECTED.to_string()));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     std::result::Result<std::string::String, ron::ser::Error>
   |     std::result::Result<std::string::String, _>
   |
   = note: an implementation of `std::cmp::PartialEq` might be missing for `std::result::Result<std::string::String, ron::ser::Error>`
```

How is this behaving for you? It's kinda important that the tests are able to compile 😛

Comment (kvark) [reply tests/depth_limit.rs]: yeah, we should compare error values, not strings

**Comment by Plecra** [comment]
So there were a couple things that came up while writing this that will need to be resolved:

https://github.com/ron-rs/ron/blob/92c1d5d595e5401a9567505d5c68e26afa2f43ac/src/ser/mod.rs#L178-L185

The serializer emits CRLF newlines on windows, but the tests don't expect them, meaning the test quite can only be run on linux. This isn't going to be hard to solve ^^

https://github.com/ron-rs/ron/blob/92c1d5d595e5401a9567505d5c68e26afa2f43ac/tests/depth_limit.rs#L58

At some point this test was broken, presumably by a `PartialEq` implementation for `ron::Error` going missing. Hopefully that can be added?

**Comment by kvark** [comment]
I think this looks good now! Thank you!
Would you mind squashing everything up to the `rustfmt` commit?
Also, would you mind rebasing instead of introducing the merge commit at the end?

**Comment by Plecra** [comment]
Ah, I just remembered! I *think* we can avoid this being a breaking change by using `String` as a default for the generic parameter, and introducing a new constructor for generic writers. Either that, or we'll need to bump the version in `Cargo.toml`

**Comment by kvark** [comment]
Yes, I agree, let's not try too hard to make this backwards compatible.

**Comment by kvark** [comment]
The CI isn't happy - https://travis-ci.org/github/ron-rs/ron

**Comment by Plecra** [comment]
It is not. I should've realized this was actually caused by my changes - using an `io::Write` meant including `io::Error` in the error type. Since it can't be `Clone`d [or `PartialEq`'d](https://github.com/rust-lang/rust/pull/34192#issuecomment-227477665), neither can `ron::Error`.

I've remade the original fix to the test.

**Comment by Plecra** [comment]
Ah shoot, how do I stop bors from merging this? Clippy's caught a pretty bad bug in my code

**Comment by kvark** [comment]
Is this all ready? Could you rebase please?

**Comment by Plecra** [comment]
Updated!

**Comment by nnethercote** [comment]
I've been using [cargo-llvm-lines](https://github.com/dtolnay/cargo-llvm-lines/) to look at compile times of webrender, in particular when the `capture` and `replay` features are on.

There was a 13% increase in the number of lines of LLVM IR generated with the `capture` feature enabled in webrender when going from `ron` 0.5.1 to 0.6.0. I'm pretty certain that this PR was the cause. This will have caused compile time regressions for webrender and other programs that use `ron`.

Here's some of the output from cargo-llvm-lines with 0.5.1:
```
  Lines        Copies         Function name
  -----        ------         -------------
1680955 (100%)  49965 (100%)  (TOTAL)
  75640 (4.5%)    620 (1.2%)  <&mut ron::ser::Serializer as serde::ser::SerializeStruct>::serialize_field
  28888 (1.7%)     92 (0.2%)  hashbrown::raw::RawTable<T>::rehash_in_place
  27980 (1.7%)   2771 (5.5%)  core::ptr::drop_in_place
  26712 (1.6%)    168 (0.3%)  hashbrown::raw::RawTable<T>::find
  23686 (1.4%)    409 (0.8%)  core::result::Result<T,E>::unwrap_or_else
  22344 (1.3%)    133 (0.3%)  <&mut ron::ser::Serializer as serde::ser::SerializeSeq>::serialize_element
  22091 (1.3%)    387 (0.8%)  core::option::Option<T>::map
  19084 (1.1%)    335 (0.7%)  core::option::Option<T>::ok_or_else
  18548 (1.1%)    128 (0.3%)  serde::ser::Serializer::collect_seq
  17296 (1.0%)     92 (0.2%)  hashbrown::raw::RawTable<T>::resize
  16893 (1.0%)    370 (0.7%)  core::ptr::swap_nonoverlapping_one
```
And here's the corresponding output with 0.6.0:
```
  Lines        Copies         Function name
  -----        ------         -------------
1897940 (100%)  50067 (100%)  (TOTAL)
 221960 (11.7%)    620 (1.2%)  <ron::ser::Compound<W> as serde::ser::SerializeStruct>::serialize_field
  40299 (2.1%)    133 (0.3%)  <ron::ser::Compound<W> as serde::ser::SerializeSeq>::serialize_element
  28888 (1.5%)     92 (0.2%)  hashbrown::raw::RawTable<T>::rehash_in_place
  28032 (1.5%)   2776 (5.5%)  core::ptr::drop_in_place
  26712 (1.4%)    168 (0.3%)  hashbrown::raw::RawTable<T>::find
  23686 (1.2%)    409 (0.8%)  core::result::Result<T,E>::unwrap_or_else
  22197 (1.2%)    389 (0.8%)  core::option::Option<T>::map
  20746 (1.1%)     82 (0.2%)  <ron::ser::Compound<W> as serde::ser::SerializeTuple>::serialize_element
  20724 (1.1%)    128 (0.3%)  serde::ser::Serializer::collect_seq
  19084 (1.0%)    335 (0.7%)  core::option::Option<T>::ok_or_else
  17296 (0.9%)     92 (0.2%)  hashbrown::raw::RawTable<T>::resize
  17052 (0.9%)     98 (0.2%)  <&mut ron::ser::Serializer<W> as serde::ser::Serializer>::serialize_newtype_variant
  16893 (0.9%)    370 (0.7%)  core::ptr::swap_nonoverlapping_one
```
This PR added the `Compound` type and logic for avoiding trailing commas, both of which increased the size of methods like `serialize_field` and `serialize_element`. These functions get instantiated 100s of times. I strongly suspect this PR is responsible for most or all of the increase.

If you want to try this yourself:
- Install cargo-llvm-lines
- Get a `webrender` clone.
- Change the `default` features line in `webrender/webrender/Cargo.toml` to `["capture"]`.
- Run `cargo +nightly llvm-lines` within `webrender/webrender`.

The non-obvious lesson here is that these methods can be instantiated very often, and so adding features/complexity can have surprisingly high costs.

**Comment by Plecra** [comment]
Reviewing the changes, I think the regressions may be primarily due to the error handling and `write_all` inlining. It's probably not specializing it for `String`s - we might be able to use a trait similar to `serde_json`'s to force it to use a simpler implementation.

Doing that would also remove the awkward `String::from_utf8` call.

**Comment by Plecra** [comment]
[`cargo-llvm-lines` is misbehaving on Windows](https://github.com/dtolnay/cargo-llvm-lines/issues/18), so I can't see the impact of the changes - Could you see what the output is like with this branch? https://github.com/Plecra/ron/tree/io-write

**Comment by nnethercote** [comment]
@Plecra: LLVM IR generation occurs before any inlining happens, so inlining isn't relevant. As I said above, I think the introduction of `Compound` and the extra trailing comma logic accounts for most of the change.

**Comment by panstromek** [comment]
@Plecra I added a workaround for the bug in `cargo-llvm-lines` you mentioned, so it should now work for Windows (just some error is logged, but you get the output).
